// @myriaddreamin/typst.ts/dist/esm/main.mjs
import * as initOptions from './options.init.mjs';
export { preloadRemoteFonts, preloadSystemFonts } from './options.init.mjs';
import * as renderer from './renderer.mjs';
export { rendererBuildInfo, createTypstRenderer, createTypstSvgRenderer } from './renderer.mjs';
import { RenderView } from './render/canvas/view.mjs';
import * as compiler from './compiler.mjs';
import { FetchAccessModel, MemoryAccessModel } from './fs/index.mjs';
import { FetchPackageRegistry } from './fs/package.mjs';
export { FetchAccessModel } from './fs/index.mjs';
export { FetchPackageRegistry } from './fs/package.mjs';
export { createTypstCompiler } from './compiler.mjs';
// Export module on window.
if (typeof window !== 'undefined') {
    window.TypstRenderModule = {
        RenderView,
        createTypstRenderer: renderer.createTypstRenderer,
        createTypstSvgRenderer: renderer.createTypstRenderer,
        preloadRemoteFonts: initOptions.preloadRemoteFonts,
        preloadSystemFonts: initOptions.preloadSystemFonts,
    };
    window.TypstCompileModule = {
        createTypstCompiler: compiler.createTypstCompiler,
        preloadRemoteFonts: initOptions.preloadRemoteFonts,
        preloadSystemFonts: initOptions.preloadSystemFonts,
        FetchAccessModel,
        MemoryAccessModel,
        FetchPackageRegistry,
        withAccessModel: initOptions.withAccessModel,
        withPackageRegistry: initOptions.withPackageRegistry,
    };
}
